libzeth/circuits/mimc/mimc_permutation.tcc

```diff
@@ -24,6 +24,8 @@ MiMC_permutation_gadget<FieldT, Exponent, NumRounds>::MiMC_permutation_gadget(
     const libsnark::pb_linear_combination<FieldT> &msg,
     const libsnark::pb_linear_combination<FieldT> &key,
     const libsnark::pb_variable<FieldT> &result,
+    const libsnark::pb_linear_combination<FieldT> &add_to_result,
+    const bool have_add_to_result,
```

Comment:
`has` would be 3rd person so seems to refer to some other class, or some other member - it's not clear.  I think `have` is more suitable, but renamed to `add_to_result_is_valid` which should leave no room for confusion.

Reply:
I see. No big deal here, the code can be followed fairly easily, but (unless I miss something)  `is_valid` isn't quite what we want, what we want is to "raise/set a flag", so "valid" may be changed to `set`? i.e. `add_to_result_is_valid` -> `add_to_result_is_set`?

Reply:
When the flag is false, the `pb_linear_combination` is completely uninitialized (and unallocated, with no terms, and certain uses of it will cause a crash if I've understood the code correctly).  It's possible for a lin comb to be allocated, but have no terms (or have terms with 0 coefficients).  If we could use this state, we could remove the flag and just treat the lin comb generically, but AFAICT it would then need to be "allocated" on the pb.  The term "unset" would seem most suited to this case of "valid" (i.e. allocated, so nothing will crash) but with no terms (i.e. does not impact any other operations).  Since that requires extra space on the PB (and, maybe sensibly, can't be done from within a constructor), that state never really occurs unless the user explicity creates and passes in the linear comb - in which case the flag is true.

Hence "valid" may be appropriate here, even though we could alternatively remove the flag and introduce some set / unset state on the lin comb.

Reply:
The code looks like this (taken from last diffs):

```cpp
    if (add_to_result_is_valid) {
        libsnark::pb_linear_combination<FieldT> key_plus_add_to_result;
        key_plus_add_to_result.assign(this->pb, key + add_to_result);
        round_gadgets.emplace_back(
            this->pb,
            round_results[NumRounds - 2],
            key,
            round_constants[NumRounds - 1],
            round_results[NumRounds - 1],
            key_plus_add_to_result,
            FMT(this->annotation_prefix, " round[%zu]", NumRounds - 1));
    } else {
        round_gadgets.emplace_back(
            this->pb,
            round_results[NumRounds - 2],
            key,
            round_constants[NumRounds - 1],
            round_results[NumRounds - 1],
            key,
            FMT(this->annotation_prefix, " round[%zu]", NumRounds - 1));
    }
```

So the boolean `add_to_result_is_valid` just serves as a flag in the conditional to select the relevant branch. The lin comb `libsnark::pb_linear_combination<FieldT> key_plus_add_to_result;` is only local to the if case. The `add_to_result` lin comb is passed as parameter though. My point was mostly that, if we want to "add `add_to_result` to the res of the permutation" it makes more sense to me to simply "set the flag" rather than "saying it is valid". Maybe the overall signature confuses me - why pass a parameter if it is "invalid" then?

Reply:
The code you are looking at is the common private constructor, which holds the init logic shared by both public constructors (one for the vanilla case, one for "add_to_result").  This shared code must therefore have a signature which can handle both cases.  I may not have understood the confusion, so let me know if that doesn't answer the question.

(Alternatively, we could copy-paste the init code over the two public constructors, in which case no flag would be required.  It's enough logic that I didn't really want to do that.)

Reply:
>  This shared code must therefore have a signature which can handle both cases.

Indeed.
So everytime the constructor is called with an `add_to_result` param, the corresponding flag `add_to_result_is_valid` must always be `true`, no? i.e. we never pass a `add_to_result` param which is "invalid"? (I'm simply trying to understand why we talk about "valid" `add_to_result` flags here, I don't quite get the rationale.)

Reply:
> we never pass a add_to_result param which is "invalid"?

Oh actually yes... I missed this "dummy" `add_to_result` param:

```cpp
          libsnark::pb_linear_combination<FieldT>(),
          false,
```

Ok I see why you named the variable `_is_valid` and not `_is_set` -> it is always set to comply with the private constructor's signature, but when it is not needed, it is set to a "dummy" value. Got it.

Reply:
Right.  The two public constructors (for the two cases) call the common init logic in the private constructor.  The `valid` flag is set to `false` if the vanilla (add-nothing-to-result) public constructor is used.